File: backend/auth/verification.ts
import { Router } from "express";
import type { Request, Response } from "express";
import prisma from "../../lib/prisma.js";
import { verifyVerificationToken, generateVerificationToken } from "../../lib/auth.js";
import { sendVerificationEmail } from "../../lib/email.js";
import { createSession, canResend, FRONTEND_URL } from "./helpers.js";

const router = Router();

// Verify email (user clicks magic link)
router.get("/verify-email", async (req: Request, res: Response) => {
    try {
        const token = req.query.token as string | undefined;
        if (!token) {
            res.redirect(`${FRONTEND_URL}/error?msg=Missing+verification+token`);
            return;
        }

        let payload: { sub: string; email: string };
        try {
            payload = verifyVerificationToken(token);
        } catch {
            res.redirect(`${FRONTEND_URL}/error?msg=Invalid+or+expired+verification+link`);
            return;
        }

        const user = await prisma.user.findUnique({ where: { id: payload.sub } });
        if (!user) {
            res.redirect(`${FRONTEND_URL}/error?msg=User+not+found`);
            return;
        }

        if (user.is_verified) {
            const accessToken = await createSession(user.id, res, req);
            res.redirect(`${FRONTEND_URL}/dashboard?token=${accessToken}`);
            return;
        }

        await prisma.user.update({
            where: { id: payload.sub },
            data: { is_verified: true },
        });

        const accessToken = await createSession(user.id, res, req);
        res.redirect(`${FRONTEND_URL}/dashboard?token=${accessToken}`);
    } catch (err) {
        console.error("[verify-email]", err);
        res.redirect(`${FRONTEND_URL}/error?msg=Verification+failed`);
    }
});

// Resend verification email (rate limited: 3 per 30 min)
router.post("/resend-verification", async (req: Request, res: Response) => {
    try {
        const { email } = req.body as { email?: string };
        if (!email) {
            res.status(400).json({ error: "Email is required" });
            return;
        }

        if (!canResend(email)) {
            res.status(429).json({ error: "Too many requests. Try again in 30 minutes." });
            return;
        }

        const user = await prisma.user.findUnique({ where: { email } });
        if (!user || user.is_verified) {
            res.json({ status: "ok" });
            return;
        }

        const verificationToken = generateVerificationToken(user.id, email);
        await sendVerificationEmail(email, verificationToken);
        res.json({ status: "ok" });
    } catch (err) {
        console.error("[resend-verification]", err);
        res.status(500).json({ error: "Internal server error" });
    }
});

// Poll verification status
router.get("/verification-status", async (req: Request, res: Response) => {
    try {
        const email = req.query.email as string | undefined;
        if (!email) {
            res.status(400).json({ error: "Email is required" });
            return;
        }

        const user = await prisma.user.findUnique({
            where: { email },
            select: { is_verified: true },
        });

        res.json({ verified: user?.is_verified ?? false });
    } catch (err) {
        console.error("[verification-status]", err);
        res.status(500).json({ error: "Internal server error" });
    }
});

export default router;

File: backend/auth-service/app/routers/verification.py
from fastapi import APIRouter, Request, Response
from fastapi.responses import RedirectResponse
from pydantic import BaseModel, EmailStr

from app.config import FRONTEND_URL
from app.database import get_pool
from app.lib.auth import verify_verification_token, generate_verification_token
from app.lib.email import send_verification_email
from app.routers._helpers import create_session, can_resend

router = APIRouter()


# ─── GET /verify-email — User clicks magic link ─────────────────────

@router.get("/verify-email")
async def verify_email(request: Request, response: Response, token: str | None = None):
    try:
        if not token:
            return RedirectResponse(f"{FRONTEND_URL}/error?msg=Missing+verification+token")

        try:
            payload = verify_verification_token(token)
        except Exception:
            return RedirectResponse(f"{FRONTEND_URL}/error?msg=Invalid+or+expired+verification+link")

        pool = get_pool()
        user = await pool.fetchrow("SELECT id, is_verified FROM users WHERE id = $1", payload["sub"])

        if not user:
            return RedirectResponse(f"{FRONTEND_URL}/error?msg=User+not+found")

        if user["is_verified"]:
            access_token = await create_session(str(user["id"]), response, request)
            return RedirectResponse(f"{FRONTEND_URL}/dashboard?token={access_token}")

        # Mark as verified
        await pool.execute("UPDATE users SET is_verified = TRUE WHERE id = $1", payload["sub"])

        access_token = await create_session(str(user["id"]), response, request)
        redirect = RedirectResponse(f"{FRONTEND_URL}/analyser?token={access_token}")
        for header_value in response.headers.getlist("set-cookie"):
            redirect.headers.append("set-cookie", header_value)
        return redirect

    except Exception as e:
        print(f"[verify-email] {e}")
        return RedirectResponse(f"{FRONTEND_URL}/error?msg=Verification+failed")


# ─── POST /resend-verification (rate limited: 3 per 30 min) ─────────

class ResendRequest(BaseModel):
    email: EmailStr


@router.post("/resend-verification")
async def resend_verification(body: ResendRequest, response: Response):
    try:
        if not can_resend(body.email):
            response.status_code = 429
            return {"error": "Too many requests. Try again in 30 minutes."}

        pool = get_pool()
        user = await pool.fetchrow(
            "SELECT id, is_verified FROM users WHERE email = $1", body.email
        )

        # Silent success if user not found or already verified (security)
        if not user or user["is_verified"]:
            return {"status": "ok"}

        token = generate_verification_token(str(user["id"]), body.email)
        await send_verification_email(body.email, token)
        return {"status": "ok"}

    except Exception as e:
        print(f"[resend-verification] {e}")
        response.status_code = 500
        return {"error": "Internal server error"}


# ─── GET /verification-status — Poll by email ───────────────────────

@router.get("/verification-status")
async def verification_status(response: Response, email: str | None = None):
    try:
        if not email:
            response.status_code = 400
            return {"error": "Email is required"}

        pool = get_pool()
        user = await pool.fetchrow(
            "SELECT is_verified FROM users WHERE email = $1", email
        )

        return {"verified": user["is_verified"] if user else False}

    except Exception as e:
        print(f"[verification-status] {e}")
        response.status_code = 500
        return {"error": "Internal server error"}
